[package] PackageExporter visualization methods

lobstergrindset opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #61148
* **#61147**
* #61146

Basic tooling to enable users to see what is inside of a PackageExporter. Added methods: 
- `externed/interned/mocked/denied_list()`: returns list of modules which are currently in the specified category
- `relied_on_by(module_name)`: returns list of modules which rely on `module_name`
- `dependency_graph_str()`: returns string format of graph for users. Example of output:
```
digraph G {
rankdir = LR;
node [shape=box];
"<res.foo.pkl>" -> "foo";
"foo" -> "torch.package";
"foo" -> "time";
"foo" -> "sentencepiece";
"foo" -> "package_top";
}
```

Differential Revision: [D29559683](https://our.internmc.facebook.com/intern/diff/D29559683)

Commit:
[package] PackageExporter visualization methods

[ghstack-poisoned]

Commit:
Update on "[package] PackageExporter visualization methods"


Basic tooling to enable users to see what is inside of a PackageExporter. Added methods: 
- `externed/interned/mocked/denied_list()`: returns list of modules which are currently in the specified category
- `relied_on_by(module_name)`: returns list of modules which rely on `module_name`
- `dependency_graph_str()`: returns string format of graph for users. Example of output:
```
digraph G {
rankdir = LR;
node [shape=box];
"<res.foo.pkl>" -> "foo";
"foo" -> "torch.package";
"foo" -> "time";
"foo" -> "sentencepiece";
"foo" -> "package_top";
}
```

[ghstack-poisoned]

Lilyjjo commented:
@Lilyjjo has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D29559683).

suo commented on torch/package/package_exporter.py:
```diff
@@ -902,6 +902,74 @@ def _can_implicitly_extern(self, module_name: str):
+        result.sort()
+        return result
+
+    def externed_list(self) -> List[str]:
```
maybe `externed_modules` and so on?

suo commented on torch/package/package_exporter.py:
```diff
@@ -902,6 +902,74 @@ def _can_implicitly_extern(self, module_name: str):
+        """
+        return self._nodes_with_action_type(_ModuleProviderAction.DENY)
+
+    def relied_on_by(self, module_name: str) -> List[str]:
```
The "rely" language is not previously present in the API/docs—I think we should try to stay consistent with talking about "dependencies". The typical term is "reverse dependency", so maybe: `get_reverse_dependencies` or `get_rdeps`,

suo commented on torch/package/package_exporter.py:
```diff
@@ -902,6 +902,74 @@ def _can_implicitly_extern(self, module_name: str):
+        return self._nodes_with_action_type(_ModuleProviderAction.DENY)
+
+    def relied_on_by(self, module_name: str) -> List[str]:
+        """Return a list of all modules which rely on the module ``module_name``.
```
rely -> depend

suo reviewed (changes requested):
Overall looks good to me, had some small comments to improve API consistency

Commit:
Update on "[package] PackageExporter visualization methods"


Basic tooling to enable users to see what is inside of a PackageExporter. Added methods: 
- `externed/interned/mocked/denied_list()`: returns list of modules which are currently in the specified category
- `relied_on_by(module_name)`: returns list of modules which rely on `module_name`
- `dependency_graph_str()`: returns string format of graph for users. Example of output:
```
digraph G {
rankdir = LR;
node [shape=box];
"<res.foo.pkl>" -> "foo";
"foo" -> "torch.package";
"foo" -> "time";
"foo" -> "sentencepiece";
"foo" -> "package_top";
}
```

Differential Revision: [D29559683](https://our.internmc.facebook.com/intern/diff/D29559683)

[ghstack-poisoned]

Lilyjjo commented:
@Lilyjjo has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D29559683).
